src/clj/harja/kyselyt/tyomaapaivakirja.sql

```diff
@@ -9,12 +9,90 @@ SELECT
+
+
+-- name: hae-paivakirjalistaus
+SELECT t.id as tyomaapaivakirja_id, t.urakka_id, u.nimi as "urakka-nimi",
+       d::DATE as paivamaara, -- Otetaan generoinnin päivämäärä
+       -- Ihan hihasta vedetty tilan määritelmä.
+       CASE
+           WHEN t.luotu IS NULL  AND d::DATE < NOW() THEN 'puuttuu'
+           WHEN t.luotu BETWEEN d::DATE and d::DATE + interval '14 day' THEN 'ok'
+           WHEN t.luotu > d::DATE + interval '14 day' THEN 'myohassa'
+       END as tila,
+       count(tk.id) as "kommenttien-maara",
+       t.versio,
+       t.luotu, t.luoja, t.muokattu, t.muokkaaja
+  FROM generate_series(:alkuaika::DATE, :loppuaika::DATE, '1 day'::interval) d
+       LEFT JOIN tyomaapaivakirja t ON t.paivamaara = d::DATE AND t.urakka_id = :urakka-id
+       JOIN urakka u ON u.id = :urakka-id
+       LEFT JOIN tyomaapaivakirja_kommentti tk on t.id = tk.tyomaapaivakirja_id
+       LEFT JOIN (SELECT versio, tyomaapaivakirja_id FROM tyomaapaivakirja_kalusto ORDER BY versio DESC limit 1) t_kalusto on t.id = t_kalusto.tyomaapaivakirja_id
+ GROUP BY t.id, u.nimi, d, t_kalusto.versio
+ ORDER BY paivamaara ASC;
+
+
+-- name: hae-paivakirja
+SELECT t.id as tyomaapaivakirja_id, t.urakka_id, u.nimi as "urakka-nimi", t.versio, t.paivamaara::DATE,
+       -- Hihasta vedetty tilan määritelmä
+       CASE
+           WHEN t.luotu BETWEEN t.paivamaara and t.paivamaara + interval '14 day' THEN 'ok'
+           WHEN t.luotu > t.paivamaara + interval '14 day' THEN 'myohassa'
+           END as tila,
+       (SELECT array_agg(row(aloitus, lopetus, nimi))
+        FROM tyomaapaivakirja_paivystaja
+        WHERE versio = :versio AND tyomaapaivakirja_id = :tyomaapaivakirja_id) as paivystajat,
+       (SELECT array_agg(row(aloitus, lopetus, nimi))
+        FROM tyomaapaivakirja_tyonjohtaja
+        WHERE versio = :versio AND tyomaapaivakirja_id = :tyomaapaivakirja_id) as tyonjohtajat,
+       (SELECT array_agg(row(havaintoaika, aseman_tunniste, aseman_tietojen_paivityshetki, ilman_lampotila, tien_lampotila, keskituuli, sateen_olomuoto, sadesumma))
+        FROM tyomaapaivakirja_saaasema
+        WHERE versio = :versio AND tyomaapaivakirja_id = :tyomaapaivakirja_id)  as "saa-asemat",
+       (SELECT array_agg(row(havaintoaika, paikka, kuvaus))
+        FROM tyomaapaivakirja_poikkeussaa
+        WHERE versio = :versio AND tyomaapaivakirja_id = :tyomaapaivakirja_id) as poikkeussaat,
+       (SELECT array_agg(row(aloitus, lopetus, tyokoneiden_lkm, lisakaluston_lkm))
+        FROM tyomaapaivakirja_kalusto
+        WHERE versio = :versio AND tyomaapaivakirja_id = :tyomaapaivakirja_id) as kalustot,
+       (SELECT array_agg(row(tyyppi::TEXT, kuvaus))
+        FROM tyomaapaivakirja_tapahtuma
+        WHERE versio = :versio AND tyomaapaivakirja_id = :tyomaapaivakirja_id) as tapahtumat,
+       (SELECT array_agg(row(kuvaus, aika))
+        FROM tyomaapaivakirja_toimeksianto
+        WHERE versio = :versio AND tyomaapaivakirja_id = :tyomaapaivakirja_id) as toimeksiannot,
```

Comment:
Jäin jo silloin koodikahveilla aikanaan miettimään mutta unohdin kommentoida: toimisiko nuo subselectit näin: 
AND tyomaapaivakirja_id = :tyomaapaivakirja_id --> 
AND tyomaapaivakirja_id = t.id

Omaan silmään se näyttäis hivenen siistimmältä ja vähemmän fragiililta.

Reply:
Näyttääpi toimivan. Vaihdan. 👍

Reply:
Tässä näyttäisi myös toimivan left join + array_agg, esim.

array_agg(row(tp.aloitus, tp.lopetus, tp.nimi)) as paivystajat,
...
  FROM tyomaapaivakirja t
         JOIN urakka u ON t.urakka_id = u.id
LEFT JOIN tyomaapaivakirja_paivystaja tp on t.id = tp.tyomaapaivakirja_id
Ainakin yksittäinen rivi miellyttää enemmän omaa silmää, mutta en sitten osaa sanoa onko tällä jotain suorituskyky tms. vaikutuksia.

Reply:
LEFT JOINilla lähdin liikkeelle. Se tuotaa useita rivejä per työmaapäiväkirja. Eli jos haluaa yhden työmaapäiväkirjan, jolla  on 5 riviä kalustoja ja vaikka 10 riviä tapahtumia, niin kahden LEFT JOIN:nin käytöllä (kalustot, tapahtumat) saadaan 15 riviä, vaikka tuloksiin yrittäisikin array_agg(row tyyppisesti niputtaa ne vastaukset.